### DISCUSSION TITLE
Using the pyControl API to interface with the PC

### DISCUSSION THREAD

**Discussion by guykoy123** [pyControl/code#124] [Q&A] [answered]
Hello everyone,
I'm am working on rather experiments so my task files have gotten quite long. I've hit a problem where I am unable to add much code because the board runs out of memory when loading the task on to it,
My problem is with a variable validation function that i want to add to make sure they are correctly set before running the task.
I want to use the API to offload that validation to the PC however i have not been able to understand how to do it from the example file.
Do I manually run a script on the PC that listens to a port or is there a way for pyControl to manage it?
Or maybe I can skip using an API and add memory to the board?
Thanks!

**Comment by alustig3** [comment]
- make sure you are using pyControl v2
- you will create a python script in the `api_classes/` directory. This script will run be run on your computer when the task in running. At a minimum it will need to have a `def process_data_user(self, data):` function where it can listen for state changes, events, prints, and variable changes coming from  your task. It can then validate based based on information is received and then talk back to the task by updating a variable or triggering an event. See the example file [here](https://github.com/pyControl/code/blob/master/api_classes/Example_user_class.py)
- you will need to add a `v.api_class` variable to your task file. See the example [here](https://github.com/pyControl/code/blob/3d1ed1efd5b91cb45646a39dc16dde1f47b829de/tasks/example/api.py#L23) 
	- (the comment on line 24 says to put the file in `config/user_classes/` but that is out of date, your api script should go in the top level folder `api_classes/`)

If you are willing to share your code and a little more explanation about what you are wanting to validate, I am happy help directly get you started (you can post python code in these discussions, instructions [here](https://github.com/orgs/pyControl/discussions/65)). 

I'd have to learn more about what you're trying to do, but I think this should be a perfect use case for the API. However if you are wanting to validate a variable that an experimenter is inputting, then a custom control dialog may be an even better solution. I can help with that as well, just let me know.

**Reply by guykoy123** [reply]
Thank you very much for the answer.
I am using pyControl v2.
I am writing a task for a behavioral experiment where a mouse is exposed to a stimulus and needs to make a decision based on that. (Different types of material will come intro contact with its' whiskers and the mouse will have to lick or wait)
To achieve that I am controlling 3 stepper motors connected to linear rails to achieve 3-axis movement.
Before running a task a lab worker will have to input positions so I want to make sure they are correct in format (list of integers in certain range) and in logic (no interference between them)
Or for example I have a probability list for each position where the sum of probabilities has to equal 1
The validation code is as follows:
```py
# check that position probabilities line up with number of positions
 # and that the sum is 1
 if len(v.position_probability_list)-1!= len(v.y_stimulus_offset):
     valid_flag=False
     print("position_probability_list need to be of length 1 + length of x_stimulus_offset list")
 if sum(v.position_probability_list)!=1:
    print("position_probability_list does not sum up to 1")
    valid_flag=False
```
Will I be able to create special function in the api class and call them directly from pyControl board or do i have to use the ```py process_data_user``` function?

**Reply by alustig3** [reply, marked as answer]
The API script can only act upon changes to the task variables after they have been made, instead we would like to prevent setting/changing the variable in the first place if it is invalid.

I have created an example for you that uses a custom controls dialog. Download and unzip [validation_v1.zip](https://github.com/user-attachments/files/16253741/validation_v1.zip). Place the `input_validation.py` in your `tasks/` folder and place `validation_gui.py` in your `controls_dialogs/` folder. 

Hope that is enough to get you started. Let me know if you have questions!



https://github.com/user-attachments/assets/72bfb217-ec44-4beb-ba68-7c131fa3ddc6

**Reply by guykoy123** [reply]
I've tried what you suggested but I'm constantly getting an AttributeError when I try to load the task
Can't seem to find the source in the validation_gui file, I just removed your variables and put a different variable, or even with no variable to add
Here is my code for Controls_gui
```py
class Controls_gui(QtWidgets.QWidget):
        def __init__(self, parent, grid_layout, board, init_vars):
            super(QtWidgets.QWidget, self).__init__(parent)
            self.board = board
            print("heeloo")
            # create widgets
            widget = QtWidgets.QWidget()
            controls_layout = QtWidgets.QGridLayout()
            row = 0

            controls_layout.addWidget(QtWidgets.QLabel("<hr>"), row, 0, 1, 4)
            row += 1

            controls_layout.setRowStretch(row, 1)
            widget.setLayout(controls_layout)
            grid_layout.addWidget(widget, 0, 0, QtCore.Qt.AlignmentFlag.AlignLeft)

            # add close shortcut
            self.close_shortcut = QtGui.QShortcut(QtGui.QKeySequence("Ctrl+W"), self)
            self.close_shortcut.activated.connect(self.close)
```
I also need to add that I moved from v2.0.1 to v2.0.2

**Reply by alustig3** [reply]
Were you able to run the files example `input_validation.py` file at all before making changes? I want to make sure we're starting from common code that works on both our computers, then we can debug what has changed. 

Assuming you were able to get the unedited example running, could you share your edited task file and validation_gui file so I can take a closer look/try to replicate the error? (you can paste the entire contents as code blocks, or change the file extenstions from .py to .txt and then attach them to this discussion, or put them and folder and zip it up and attach the zipped folder to this discussion)

**Reply by alustig3** [reply]
also please check/share your `ErrorLog.txt` file as there may be more information about the AttributeError there. You can view the log by going to the menubar and selecting `View->Error log`. It should also exist as a .txt file in the top level of your pyControl directory that you are running the program from.

**Comment by guykoy123** [comment]
The code you provided worked great, attached is my code for simple task that spins a mouse wheel and the validation i tried to create based on your code.
[T1_wheel_moving.zip](https://github.com/user-attachments/files/16506560/T1_wheel_moving.zip)

This is the error i pulled from the `ErrorLog.txt`
```2024-08-06 09:58:52,827 Traceback (most recent call last):
  File "C:\Users\guykoy\Desktop\pyControl - development\Physiology_Lab_BGU\source\gui\run_task_tab.py", line 387, in setup_task
    self.controls_dialog = py_gui_file.Custom_controls_dialog(self, self.board)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\guykoy\Desktop\pyControl - development\Physiology_Lab_BGU\controls_dialogs\T1_validation_gui.py", line 20, in __init__
    self.controls_grid = Controls_grid(self, board)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\guykoy\Desktop\pyControl - development\Physiology_Lab_BGU\controls_dialogs\T1_validation_gui.py", line 32, in __init__
    self.controls_gui = Controls_gui(self, self.grid_layout, board, initial_variables_dict)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\guykoy\Desktop\pyControl - development\Physiology_Lab_BGU\controls_dialogs\T1_validation_gui.py", line 45, in __init__
    Spin_var()
TypeError: Spin_var.__init__() missing 6 required positional arguments: 'init_var_dict', 'label', 'spin_min', 'spin_max', 'step', and 'varname'
```



Even though i made sure all the variable are sent when calling the function.

**Reply by alustig3** [reply]
I believe the mistake is on lines 45 and 46 of the `T1_validation_gui.py` that you attached. Currently they are:

```python
Spin_var()
self.number_of_trials=Text_var(init_var_dict=init_vars,label="<b>Number of trials:</b>",varname="number_of_trials",stpe=1,spin_min=-1,spin_max=400 ,text_width=INPUT_WIDTH)
```
delete or comment out line 45 and edit line 46 to use `Spin_var` instead of `Text_var`. So it should look like:
```python
# Spin_var() comment or delete this line
self.number_of_trials = Spin_var(
    init_var_dict=init_vars,
    label="<b>Number of trials:</b>",
    varname="number_of_trials",
    stpe=1,
    spin_min=-1,
    spin_max=400,
    text_width=INPUT_WIDTH,
)
```

**Reply by guykoy123** [reply]
That, plus removing a few more unecessary lines fixed the problem, can't believe I missed it.
Thanks you so much!

**Reply by alustig3** [reply]
You're welcome, happy to hear it is working!

**Reply by guykoy123** [reply]
Small question about the validating function
In this part:
```py
             # connect the validator to the line edit and set button
            self.probabilites.line_edit.returnPressed.disconnect()  # disconnect the default returnPressed event
            self.probabilites.set_btn.clicked.disconnect()  # disconnect the default clicked event
            self.probabilites.line_edit.returnPressed.connect(probabilities_validator)
           self.probabilites.set_btn.clicked.connect(probabilities_validator)
```
there is no `line_edit` property for a Spin_Var, what do I do to disconnect it?

**Reply by guykoy123** [reply]
Is it also possible to silence the message windows that pop up for variable changed?

**Reply by alustig3** [reply]
> there is no line_edit property for a Spin_Var, what do I do to disconnect it?


The `Spin_var` will automatically restrict the input to be in between your `spin_min` and `spin_max`, as a form of basic of validation. If you would like to add additional validation, you can create a validation function then connect it to the `Spin_var`'s `set_btn`

here is an example: 
```py
    class Controls_gui(QtWidgets.QWidget):
        def __init__(self, parent, grid_layout, board, init_vars):
            super(QtWidgets.QWidget, self).__init__(parent)
            self.board = board

            # create widgets
            widget = QtWidgets.QWidget()
            controls_layout = QtWidgets.QGridLayout()
            row = 0
            self.number_of_trials = Spin_var(
                init_var_dict=init_vars,
                label="<b>Number of trials:</b>",
                varname="number_of_trials",
                step=1,
                spin_min=-1,
                spin_max=400,
            )
            self.number_of_trials.add_to_grid(controls_layout, row)
            self.number_of_trials.setBoard(board)

            # create a validator function for the spin_var
            def multiple_of_3_validator():
                # validate that the input is a multiple of 3
                if self.number_of_trials.spn.value() % 3 != 0:
                    msg = QtWidgets.QMessageBox()
                    msg.setText("Number of trials should be a mulitple of 3")
                    msg.exec()
                    return
                self.number_of_trials.set()

            # connect the validator to the spin_var's set button
            self.number_of_trials.set_btn.clicked.disconnect()  # disconnect the default clicked event
            self.number_of_trials.set_btn.clicked.connect(multiple_of_3_validator)
            row += 1
```

> Is it also possible to silence the message windows that pop up for variable changed?

that popup will only appear when you change variables before starting the task. During the task, changes to the variable will shown in / recorded to the data log. If you want to suppress the message you can comment out this line: https://github.com/pyControl/code/blob/3d1ed1efd5b91cb45646a39dc16dde1f47b829de/source/gui/custom_controls_dialog.py#L78

**Reply by guykoy123** [reply]
Thank you so much for everything!!!
